posts/2023/08/03/ansible-data-manipulation-with-a-filter.md

```diff
@@ -0,0 +1,99 @@
+---
+title: Ansible data manipulation with a Filter
+author: Sean Sullivan
+date: 2023-08-03 13:00:00
+slug: ansible-data-manipulation-with-a-filter
+category: developer
+tags: developer, filters, playbook
+type: text
+---
+
+
+## Background:
+
+This year at Summit, an attendee posed a question about how to work with setting facts and changing data in Ansible. Many times we’ve come across people using task after task to manipulate data, to turn items into lists, filter our options, trying to do heavy data manipulation and to turn data from one source into another. Trying to make these programmatic changes using a mixture of YAML and Jinja inside of roles and playbooks is a headache of its own. While many of these options will work, they aren’t very efficient or easy to implement. Ansible Playbooks were never meant for programming.
+
+One solution that is usually overlooked is to do the manipulation in Python inside of a module or a filter. This article will detail how to create a filter to manipulate data. In addition, [a repository for all code referenced](https://github.com/sean-m-sullivan/ansible_data_manipulation_examples) in this article has been created. 
+
+This example was first developed as a module. However after review, it was determined that these data transformations are best done as filters. Filters can take multiple data inputs, do the programmatic operations, and then can be used in line where they are used as input or set as a fact. In addition, this runs locally and not at the host level, so it can be faster and avoid unnecessary connections.
+
+## Starting Point
+
+To begin, we need a dataset to work on. For this we used data from the automation controller API, workflows; it gives nested data on the nodes in each workflow to loop around. The variable file used in this case [can be found in the repo](https://github.com/sean-m-sullivan/ansible_data_manipulation_examples/blob/main/workflows.yml).
+
+The goal is to find what is being used in the automation controller looping over nested lists. While this is not a very practical example, it does give a starting point for creating a filter to manipulate any dataset.
+
+## Filter Basics
+
+The bones of this filter were taken from [ansible.netcommon.pop_ace](https://github.com/ansible-collections/ansible.netcommon/blob/main/plugins/filter/pop_ace.py). The start of every filter has some required options, such as FilterModule, and in addition AnsibleFilterError is good for troubleshooting.
+
+    from ansible.errors import AnsibleFilterError
```

Comment:
Looks like the code block here is from spaces. It works, but it's not consistent with what we have in the other submissions (backticks or ~~~), not sure if it's a problem, but highlighting just in case.